core/src/main/java/com/linecorp/armeria/common/logging/PrometheusMetricRequestDecorator.java

```diff
@@ -0,0 +1,351 @@
+import com.linecorp.armeria.server.Service;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.prometheus.client.Collector;
+import io.prometheus.client.CollectorRegistry;
+import io.prometheus.client.Counter;
+import io.prometheus.client.Gauge;
+import io.prometheus.client.Summary;
+
+/**
+ * Provides request decorator for server and client sides. Decorators will record metrics into
+ * Prometheus {@link CollectorRegistry}.
+ * @param <I> {@link Request} type
+ * @param <O> {@link Response} type
+ */
+public final class PrometheusMetricRequestDecorator<I extends Request, O extends Response>
+        implements DecoratingServiceFunction<I, O>, DecoratingClientFunction<I, O> {
+
+    private static class MetricAdapter {
+        private AtomicReference<MetricAdapter> initialized;
+
+        private Summary timer;
+        private Counter success;
+        private Counter failure;
+        private Gauge activeRequests;
+        private Summary requestBytes;
+        private Summary responseBytes;
+
+        private List<Method> metricLabelValueMethods;
+
+        MetricAdapter() {
+            initialized = new AtomicReference<>();
+        }
+
+        public MetricAdapter initialize(MetricLabel<?> namingFunction,
+                                        PrometheusRegistryWrapper collectorRegistry) {
+            if (initialized.compareAndSet(null, this)) {
+                BeanInfo metricLabelValueBeanInfo;
+
+                try {
+                    metricLabelValueBeanInfo = Introspector.getBeanInfo(namingFunction.getValue().getClass());
+                } catch (IntrospectionException e) {
+                    throw new RuntimeException(e);
+                }
+
+                String[] labels = Arrays.asList(metricLabelValueBeanInfo.getPropertyDescriptors())
+                                        .stream()
+                                        .skip(1)
+                                        .map(PropertyDescriptor::getName)
+                                        .map(Collector::sanitizeMetricName)
+                                        .toArray(String[]::new);
+
+                String sanitizedName = namingFunction.getTokenizedName()
+                                                     .stream()
+                                                     .collect(Collectors.joining("_"));
+
+                metricLabelValueMethods = Arrays.asList(metricLabelValueBeanInfo.getPropertyDescriptors())
+                                                .stream()
+                                                .skip(1)
+                                                .map(PropertyDescriptor::getReadMethod)
+                                                .collect(Collectors.toList());
+
+                timer = create(sanitizedName + "_timer",
+                               collectorRegistry,
+                               name -> Summary.build()
+                                              .quantile(.5, .01)
```

Comment:
Do you think it would be better to make this configurable? basically I copied what Dropwizard has since we were exporting metrics from there originally.